crates/ruff/src/rules/perflint/rules/incorrect_dict_iterator.rs

```diff
@@ -142,13 +142,33 @@ impl fmt::Display for DictSubset {
     }
 }
 
-/// Returns `true` if the given expression is either an ignored value or a tuple of ignored values.
-fn is_ignored_tuple_or_name(expr: &Expr, dummy_variable_rgx: &Regex) -> bool {
+/// Returns `true` if the given expression is either an unused value or a tuple of unused values.
+fn is_unused(expr: &Expr, model: &SemanticModel) -> bool {
     match expr {
-        Expr::Tuple(ast::ExprTuple { elts, .. }) => elts
-            .iter()
-            .all(|expr| is_ignored_tuple_or_name(expr, dummy_variable_rgx)),
-        Expr::Name(ast::ExprName { id, .. }) => dummy_variable_rgx.is_match(id.as_str()),
+        Expr::Tuple(ast::ExprTuple { elts, .. }) => elts.iter().all(|expr| is_unused(expr, model)),
+        Expr::Name(ast::ExprName { id, .. }) => {
+            // Treat a variable as used if it has any usages, _or_ it's shadowed by another variable
+            // with usages.
+            //
+            // If we don't respect shadowing, we'll incorrectly flag `bar` as unused in:
+            // ```python
+            // from random import random
+            //
+            // for bar in range(10):
+            //     if random() > 0.5:
+            //         break
+            // else:
+            //     bar = 1
+            //
+            // print(bar)
+            // ```
+            let scope = model.scope();
+            scope
+                .get_all(id)
+                .map(|binding_id| model.binding(binding_id))
+                .filter(|binding| binding.range.start() >= expr.range().start())
+                .all(|binding| !binding.is_used())
```

Comment:
This does have some weird false negatives, like:

```python
def f():
    # Unfixable (false negative) due to usage of `bar` outside of loop.
    for foo, bar in dict.items():
        break

    bar = 1
    print(bar)
```

We'd consider `bar` used, because it's shadowed by a future `bar`, and that shadowed variable is used. We can fix this with better control flow logic, but it's fine for now, `B007` already suffers from this.